Move Photon to a separate package - "Image CDN"

dilirity opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Add Image CDN package;
* Add Image CDN as a separate optimization module to Jetpack Boost.
* Updated Boost compatibility file to sync status between image-cdn in boost and photon in jetpack.
* Disable photon in jetpack when image-cdn is present and use that instead

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

No.

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

No.

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Activate Image CDN and test if image URLs start with https://i0.wp.com
* Make sure **Jetpack -> Settings -> Performance -> "Speed up image load times"** syncs with **Boost -> "Image CDN"** when you toggle the status on either place
* Make sure, photon inside Jetpack does not load if Boost has Image CDN enabled. There is no visible way to verify this as the functionality is almost same between Image CDN and Photon, But, you can try hacking by adding `wp_die()` or something in the photon file in Jetpack.

> **Note**
> I have investigated the failing e2e tests. The tests seem broken because Jetpack asks to prove humanity when the e2e tests are running. The tests should be updated and in a different PR.

dilirity committed:
Add image CDN package

dilirity committed:
changelog

dilirity committed:
changelog

dilirity committed:
Add image CDN to Boost

haqadn committed:
Generate a fresh package for CDN

haqadn committed:
Update lock files

haqadn committed:
Update namespaces

haqadn committed:
Fix compatibility

dilirity committed:
Update Boost's UI to include Image CDN optimization module

dilirity committed:
Fix unable to load photon.js

haqadn commented:
More things to do:

- [ ] Configure webpack to build and minify the js file
- [ ] Remove photon code from Jetpack and depend on the plugin (separate PR)
- [ ] Update WPCOM to adapt to these changes

dilirity committed:
Remove compatibility

It doesn't belong in the module.

dilirity committed:
Remove unnecessary requires

anomiex commented on projects/packages/image-cdn/.phpcsignore:
Not needed, the listed file does not exist.

anomiex commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -0,0 +1,1390 @@
+		wp_enqueue_script(
+			'jetpack-photon',
+			Assets::get_file_url_for_environment(
+				'js/photon.min.js',
```
I note that you don't have any build steps actually creating the minified file. When you do add the build steps, I'd recommend not putting them at this location as it seems to work out cleaner when there's a `build/` or `dist/` directory for built files instead. The lazy-images package might work as an example.

Also, when you do add that build step, it'll probably be worth building in the more modern manner so you can use `Assets::register_script()` instead of this old way.

(for that matter, someday we should go through Jetpack's modules and do the same now that wpcom isn't being weird about them anymore, but that's way off-topic for this PR)

anomiex commented on projects/packages/image-cdn/src/js/photon.min.js:
This should be built, not checked into the repo.

I note the copy you have here isn't actually even minimized.

anomiex commented on projects/packages/image-cdn/src/js/photon.min.js.map:
This should be built, not checked into the repo.

And for that matter, whether it should be included in the package or not is an open question. [The lazy-images package doesn't](https://github.com/Automattic/jetpack-lazy-images/tree/trunk/dist).

anomiex reviewed (commented):
Gave the package structure a once-over. Didn't review the code at all.

haqadn committed:
Disable photon functionality in Jetpack

kraftbj commented on projects/packages/image-cdn/tests/php/bootstrap.php:
```diff
@@ -0,0 +1,11 @@
+/**
+ * Include the composer autoloader.
+ */
+require_once __DIR__ . '/../../vendor/autoload.php';
```
Can we port the tests over at the same time? Photon is an area of Jetpack that does have a good amount of unit tests written 

ex: https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/tests/php/test_class.jetpack_photon.php

https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/tests/php/test_functions.photon.php

With moving/refactoring, we can lean on these to help us ensure we're keeping the functionality (or at least are aware of gaps) and prevent regressions.

haqadn committed:
Merge branch 'add/package/image-cdn' into update/photon-compatibility

haqadn committed:
Sync module status

dilirity committed:
[not verified] Remove PHPCS ignore file

haqadn committed:
[not verified] Remove phpcs ignore for nonexistent file

haqadn committed:
[not verified] Remove minified files for photon.js

The original js file is tiny already. Minifying it increases the size.
So, going with no minification or webpack configuration for now

haqadn committed:
Move core image cdn functions to a class

haqadn committed:
Load image CDN core while loading jetpack

dilirity committed:
Ignore WordPress

Commit:
Merge branch 'trunk' into add/package/image-cdn

Commit:
Fix composer deps for boost

Commit:
[not verified] Move unit tests for photon to image-cdn. Many are still broken.

thingalon commented:
@haqadn I've added the photon unit test suite, but many of the tests are failing. :) Since I'm about to board a flight, I figured I'd push what I have so that you can roll with it if you get time.

I have converted this to a draft until the unit tests are passing.

haqadn committed:
Move Status to required from required-dev to fix tests

haqadn committed:
Use WorDBless

haqadn committed:
Force `wp_attachment_image_url()` to to return absolute URL

haqadn committed:
Update user creation to support WorDBless

haqadn committed:
Fix references to `Image_CDN_Core` methods

haqadn committed:
Fix errors

haqadn committed:
Fix test cases

haqadn committed:
Rename test file

haqadn committed:
Rename test method

Commit:
Merge branch 'trunk' into add/package/image-cdn

Commit:
changelog

Commit:
Move the package version constant to the main class

haqadn committed:
Undo changes in Jetpack to keep it unmodified in this PR

haqadn committed:
Fix project versions

haqadn committed:
Fix phpcs

thingalon commented:
Commenting here to document a discussion and decision around minifying `photon.js`:
- `photon.js` is 1.72 KiB unminified.
- Once Webpack is done with it, `photon.min.js` ends up being 🥁 2.03 KiB large.
![image](https://user-images.githubusercontent.com/1369626/228240085-d63c7569-5109-47e1-bb2e-27bee7088301.png)

So, we decided *not* to add a webpack step to the package. It may be worth minifying photon.js in a more efficient manner later, but for now not minifying it is a win.

thingalon reviewed (dismissed):
This is looking great.

I have managed to replicate the Boost E2E test errors _intermittently_ on trunk. I think we have more work to do there to make sure our E2E tests don't belch errors like this, but I don't believe that should hold up this PR.

haqadn committed:
Merge branch 'update/photon-compatibility' into add/package/image-cdn

haqadn committed:
Elaborate comment about photon compatibliity

thingalon reviewed (dismissed):
Thanks for merging in the compatibility module. This is working well, and I think it's ready to merge :)

thingalon commented:
@Automattic/jetpack-crew - this has passed team review. Would really love an extra pair of Jetpack-experienced eyes over it before we merge :)

anomiex commented:
>     * Once Webpack is done with it, `photon.min.js` ends up being drum 2.03 KiB large.
>       ![image](https://user-images.githubusercontent.com/1369626/228240085-d63c7569-5109-47e1-bb2e-27bee7088301.png)

Is that with a production build, or a dev build that doesn't actually minify? When I look at https://github.com/Automattic/jetpack-production/blob/trunk/_inc/build/photon/photon.min.js it's only 685 bytes.

OTOH, 1.72K isn't a huge amount anyway. :shrug:

anomiex commented on projects/packages/image-cdn/src/class-image-cdn-core.php:
```diff
@@ -0,0 +1,390 @@
+	 * @see https://developer.wordpress.com/docs/photon/
+	 *
+	 * @param string       $image_url URL to the publicly accessible image you want to manipulate.
+	 * @param array|string $args An array of arguments, i.e. array( 'w' => '300', 'resize' => array( 123, 456 ) ), or in string form (w=123&h=456).
```
typo
```suggestion
	 * @param array|string $args An array of arguments, e.g. array( 'w' => '300', 'resize' => array( 123, 456 ) ), or in string form (w=123&h=456).
```

anomiex commented on projects/packages/image-cdn/src/class-image-cdn-core.php:
```diff
@@ -0,0 +1,390 @@
+				$photon_url = add_query_arg( $args, $photon_url );
+			} else {
+				// You can pass a query string for complicated requests but where you still want CDN subdomain help, etc.
+				$photon_url .= '?' . $args;
```
Looks like there's probably a bug here if line 211 was also reached.

thingalon replied:
This probable bug already exists in the [Jetpack source this package was built from](https://github.com/Automattic/jetpack/blob/1009e246529375ad28a3ed221eb02574ff24962f/projects/plugins/jetpack/functions.photon.php#L181-L192). But I agree; it looks like something we will need to fix.

I don't know how to replicate this issue, and don't want to hold this PR up for it. So I've filed [a new ticket for this issue](https://github.com/Automattic/jetpack/issues/29770).

anomiex commented on projects/packages/image-cdn/src/class-image-cdn-core.php:
```diff
@@ -0,0 +1,390 @@
+			'/\.paypalobjects\.com$/',
+			'/\.dropbox\.com$/',
+			'/\.cdninstagram\.com$/',
+			'/^(commons|upload)\.wikimedia\.org$/',
+			'/\.wikipedia\.org$/',
```
Probably not an issue for this PR, I just happened to see this and wanted to make mention of a few things here:

* `upload.wikimedia.org` should be safe to Photonize if we want, as that's the domain used to serve the actual images (e.g. https://upload.wikimedia.org/wikipedia/commons/2/20/WordPress_logo.svg or https://upload.wikimedia.org/wikipedia/commons/0/0f/Wp-twentyten.jpg).
  * Also of note is that these might include their own scaling (e.g. https://upload.wikimedia.org/wikipedia/commons/thumb/2/20/WordPress_logo.svg/100px-WordPress_logo.svg.png or https://upload.wikimedia.org/wikipedia/commons/thumb/0/0f/Wp-twentyten.jpg/220px-Wp-twentyten.jpg)
* As for the URLs like https://commons.wikimedia.org/wiki/File:WordPress_logo.svg or https://commons.wikimedia.org/wiki/File:Wp-twentyten.jpg that we're trying to avoid here, I wonder if it would be useful to match the path against `/wiki/(?:File|Image):` to handle many other MediaWiki installations as well as these (like say https://en.wikibooks.org/wiki/File:WordPress_logo.svg or https://logos.fandom.com/wiki/File:Wordpress_logo.svg).

cc @jeherve

jeherve replied:
You know the Wiki infra a lot more than me, so I'll defer to you on that :)

> I wonder if it would be useful to match the path against /wiki/(?:File|Image):

This makes sense to me. it seems it would catch all the examples you mentioned; I don't know if there are any other URL formats that we need to be catching.

anomiex replied:
The main other "format" we might care about catching is that the "File" and "Image" strings could be localized (and in any case).

And while `/wiki/` should work for Wikimedia and Fandom, there's nothing requiring it. Even some Wikimedia sites might use something different in some cases, like https://zh.wikipedia.org/zh-tw/File:WordPress_logo.svg. But there's really not much we can sensibly do about that.

anomiex commented on projects/packages/image-cdn/src/class-image-cdn-core.php:
```diff
@@ -0,0 +1,390 @@
+	public static function get_jetpack_content_width() {
+		$content_width = ( isset( $GLOBALS['content_width'] ) && is_numeric( $GLOBALS['content_width'] ) )
+		? $GLOBALS['content_width']
+		: false;
```
Nitpick
```suggestion
		$content_width = ( isset( $GLOBALS['content_width'] ) && is_numeric( $GLOBALS['content_width'] ) )
			? $GLOBALS['content_width']
			: false;
```

anomiex commented on projects/packages/image-cdn/src/class-image-cdn-image-sizes.php:
```diff
@@ -0,0 +1,191 @@
+	 */
+	protected function generate_sizes() {
+
+		// There is no need to generate the sizes a new for every single image.
```
typo
```suggestion
		// There is no need to generate the sizes anew for every single image.
```

anomiex commented on projects/packages/image-cdn/src/class-image-cdn-image-sizes.php:
```diff
@@ -0,0 +1,191 @@
+		/** This filter is documented in wp-admin/includes/image.php */
+		$sizes = apply_filters( 'intermediate_image_sizes_advanced', self::$sizes, $this->data );
+
+		// Re-add the filter removed above.
```
What if the filter wasn't present in the first place, e.g. because Jetpack isn't loaded on the site?

haqadn replied:
Nice catch. It appears the intention was to remove the filter the photon module added. In this case, the package added the filter that should be removed. We missed the reference. and I have pushed an update.

anomiex replied:
Looks like that the intention is that this also depends on the code existing to disable Jetpack's copy of Photon when the package is used?

thingalon replied:
Yes, this relies on the inbuilt Jetpack-Photon being deactivated by Boost's compatibility code to prevent its `intermediate_image_sizes_advanced` hooks from getting in the way.

When `jetpack_photon_noresize_mode` is enabled (via a filter) and I hack out Boost's code to deactivate Jetpack Photon, this breaks.

Just to be extra-defensive I have committed e5d7fd1728 which removes the inbuilt-Jetpack version of this filter just-in-case.

anomiex commented on projects/packages/image-cdn/src/class-image-cdn-image-sizes.php:
```diff
@@ -0,0 +1,191 @@
+		remove_filter(
+			'intermediate_image_sizes_advanced',
+			array( 'Jetpack_Photon', 'filter_photon_noresize_intermediate_sizes' )
+		);
```
Also `array( Image_CDN::class, 'filter_photon_noresize_intermediate_sizes' )`?

And should Jetpack's version of this be updated to do the same?

haqadn replied:
Didn't understand what you meant by
> And should Jetpack's version of this be updated to do the same?

So, please let me know if I missed your point.


When this package is present(via boost), Jetpack's version will not. But, the class reference was wrong in the package, which I have updated now.

anomiex replied:
> When this package is present(via boost), Jetpack's version will not.

Ok, I see that now. I suppose there's some risk that if anyone else decides to use this package in their plugin, they won't necessarily have Boost's code that disables it in Jetpack though.

Would it simplify things to put that code into the package instead?

thingalon replied:
Moved in 3835de43d83ec79200ded9a9f02e3582800eb041.

anomiex commented on projects/packages/image-cdn/src/class-image-cdn-core.php:
```diff
@@ -0,0 +1,390 @@
+	/**
+	 * Register hooks.
+	 */
+	public static function setup_hooks() {
```
Should there be some sort of guard in here against double setup, for the future when both Boost and Jetpack use this?

For that matter, what should we do about the possibility of both this (via Boost) and Jetpack's existing Photon code both being active on a site? Will that break any of these filters? Will any of this stuff break, or is it all idempotent?

haqadn replied:
No, we tested this, and WordPress takes care of double hooking if both are on the same priority level.

anomiex commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -0,0 +1,1386 @@
+	/**
+	 * Image sizes.
+	 *
+	 * Don't access this directly. Instead, use self::image_sizes() so it's actually populated with something.
```
Should we make it private in the new class then?

thingalon replied:
Agreed. Done in 4ea3c040aa.

anomiex commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -0,0 +1,1386 @@
+	 */
+	public static function instance() {
+		if ( ! is_a( self::$instance, 'Automattic\Jetpack\Image_CDN' ) ) {
+			self::$instance = new Image_CDN();
```
Nitpick 🙂
```suggestion
		if ( ! is_a( self::$instance, self::class ) ) {
			self::$instance = new self();
```
(or `static` instead of `self`)

anomiex commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -0,0 +1,1386 @@
+	 * @uses add_action, add_filter
+	 * @return void
+	 */
+	private function setup() {
```
As above, is there any concern about all these filters running along with the `Jetpack_Photon` versions from Jetpack?

haqadn replied:
When this package is present, it will prevent the photon module from loading to avoid any potential conflict. The module will still show enabled in the UI, but this package will be active under the hood.

anomiex replied:
> When this package is present

Well, when Boost is present. See my reply to https://github.com/Automattic/jetpack/pull/29561#discussion_r1151620450.

thingalon replied:
Nice; I see @haqadn has moved the compatibility checks into the module in 3835de43d83ec79200ded9a9f02e3582800eb041.

anomiex replied:
Great, thanks!

anomiex commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -0,0 +1,1386 @@
+/**
+ * Class Image_CDN
+ */
+class Image_CDN {
```
I wonder if this should be `final`. Or else should all the `__CLASS__` stuff in here be changed to `static::class` or the like?

thingalon replied:
Going with `final` in d074d87bfe. There's no reason to inherit from it.

anomiex commented on projects/packages/image-cdn/src/class-image-cdn-core.php:
```diff
@@ -0,0 +1,390 @@
+			 *
+			 * @module photon
+			 *
+			 * @since 4.1.0
```
I wonder if all these hook docs should be updated somehow for the package. At a quick grep, I see some packages haven't been, and some use `@since-jetpack` for the Jetpack version.

thingalon replied:
What is the difference between `@since` and `@since-jetpack`?

For most of these, I believe they belong in there unmodified, as they are hooks that are not changing name.

However, in 6a1acfdfe4 I've removed some `@since` tags from methods which now exist in a class which did not previously exist (e.g.: `Image_CDN`), since they are no longer where they used to be `@since` that prior version.

anomiex replied:
I guess that `@since` refers to the package version while `@since-jetpack` refers to the Jetpack plugin version, particularly for stuff like this that was split out of Jetpack. I have no idea how/if we actually process them for https://developer.jetpack.com/ or anywhere else.

anomiex commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -0,0 +1,1386 @@
+			true === is_array( $data )
+			&& true === array_key_exists( 'sizes', $data )
+			&& true === is_array( $data['sizes'] )
+			&& false === empty( $data['sizes'] )
```
I wonder whether we should take the opportunity to clean up some of this kind of stuff. There seems no point in doing all these `true ===` and `false ===` (here and elsewhere) for boolean-returning functions like these, or `1 === preg_match()`, etc.

thingalon replied:
We're already shifting around a lot here, I'd prefer to minimise changes which don't change functionality in this move.

anomiex commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -0,0 +1,1386 @@
+	public static function parse_images_from_html( $content ) {
+		$images = array();
+
+		if ( preg_match_all( '#(?:<a[^>]+?href=["|\'](?P<link_url>[^\s]+?)["|\'][^>]*?>\s*)?(?P<img_tag><(?:img|amp-img|amp-anim)[^>]*?\s+?src=["|\'](?P<img_url>[^\s]+?)["|\'].*?>){1}(?:\s*</a>)?#is', $content, $images ) ) {
```
Pretty sure there's some typos in this regex
```suggestion
		if ( preg_match_all( '#(?:<a[^>]+?\s+?href=["\'](?P<link_url>[^\s]+?)["\'][^>]*?>\s*)?(?P<img_tag><(?:img|amp-img|amp-anim)[^>]*?\s+?src=["\'](?P<img_url>[^\s]+?)["\'].*?>){1}(?:\s*</a>)?#is', $content, $images ) ) {
```
Or do we really want to be looking for `<a href=|url|>`?

thingalon replied:
Good eye. Your suggested update requires two characters between `<a` and `href` though, so `<a href='cheese"><img src='x'></a>` doesn't match.

I've gone ahead with `[^>]*?\s+?` in be6dfd69f6 to allow 0 characters before the space before href.

anomiex replied:
Oops, good catch!

anomiex commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -0,0 +1,1386 @@
+
+				// Support Automattic's Lazy Load plugin.
+				// Can't modify $tag yet as we need unadulterated version later.
+				if ( preg_match( '#data-lazy-src=["|\'](.+?)["|\']#i', $images['img_tag'][ $index ], $lazy_load_src ) ) {
```
As above
```suggestion
				if ( preg_match( '#data-lazy-src=["\'](.+?)["\']#i', $images['img_tag'][ $index ], $lazy_load_src ) ) {
```

anomiex commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -0,0 +1,1386 @@
+					$placeholder_src      = $placeholder_src_orig;
+					$src_orig             = $lazy_load_src[1];
+					$src                  = $src_orig;
+				} elseif ( preg_match( '#data-lazy-original=["|\'](.+?)["|\']#i', $images['img_tag'][ $index ], $lazy_load_src ) ) {
```
```suggestion
				} elseif ( preg_match( '#data-lazy-original=["\'](.+?)["\']#i', $images['img_tag'][ $index ], $lazy_load_src ) ) {
```

anomiex commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -0,0 +1,1386 @@
+
+					// First, check the image tag. Note we only check for pixel sizes now; HTML4 percentages have never been correctly
+					// supported, so we stopped pretending to support them in JP 9.1.0.
+					if ( preg_match( '#[\s|"|\']width=["|\']?([\d%]+)["|\']?#i', $images['img_tag'][ $index ], $width_string ) ) {
```
```suggestion
					if ( preg_match( '#[\s"\']width=["\']?([\d%]+)["\']?#i', $images['img_tag'][ $index ], $width_string ) ) {
```

anomiex commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -0,0 +1,1386 @@
+						$width = false === strpos( $width_string[1], '%' ) ? $width_string[1] : false;
+					}
+
+					if ( preg_match( '#[\s|"|\']height=["|\']?([\d%]+)["|\']?#i', $images['img_tag'][ $index ], $height_string ) ) {
```
```suggestion
					if ( preg_match( '#[\s"\']height=["\']?([\d%]+)["\']?#i', $images['img_tag'][ $index ], $height_string ) ) {
```

anomiex commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -0,0 +1,1386 @@
+					}
+
+					// Detect WP registered image size from HTML class.
+					if ( preg_match( '#class=["|\']?[^"\']*size-([^"\'\s]+)[^"\']*["|\']?#i', $images['img_tag'][ $index ], $size ) ) {
```
```suggestion
					if ( preg_match( '#class=["\']?[^"\']*size-([^"\'\s]+)[^"\']*["\']?#i', $images['img_tag'][ $index ], $size ) ) {
```

anomiex commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -0,0 +1,1386 @@
+
+					// WP Attachment ID, if uploaded to this site.
+					if (
+						preg_match( '#class=["|\']?[^"\']*wp-image-([\d]+)[^"\']*["|\']?#i', $images['img_tag'][ $index ], $attachment_id ) &&
```
```suggestion
						preg_match( '#class=["\']?[^"\']*wp-image-([\d]+)[^"\']*["\']?#i', $images['img_tag'][ $index ], $attachment_id ) &&
```

anomiex commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -0,0 +1,1386 @@
+						//
+						// If we are transforming the image via one of those methods, let's update the width and height attributes.
+						if ( empty( $args['resize'] ) && empty( $args['fit'] ) && empty( $args['lb'] ) ) {
+							$new_tag = preg_replace( '#(?<=\s)(width|height)=["|\']?[\d%]+["|\']?\s?#i', '', $new_tag );
```
```suggestion
							$new_tag = preg_replace( '#(?<=\s)(width|height)=["\']?[\d%]+["\']?\s?#i', '', $new_tag );
```

anomiex commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -0,0 +1,1386 @@
+							// (["|\']?)       - Matches, and captures, ", ', or empty string
+							// \s              - Ensures there's a space after the attribute
+							$new_tag = preg_replace( '#(?<=\s)(width=["|\']?)[\d%]+(["|\']?)\s?#i', sprintf( '${1}%d${2} ', $resize_args[0] ), $new_tag );
+							$new_tag = preg_replace( '#(?<=\s)(height=["|\']?)[\d%]+(["|\']?)\s?#i', sprintf( '${1}%d${2} ', $resize_args[1] ), $new_tag );
```
```suggestion
							// (?<=\s)        - Ensure width or height attribute is preceded by a space
							// (width=["\']?) - Matches, and captures, width=, width=", or width='
							// [\d%]+         - Matches 1 or more digits or percent signs
							// (["\']?)       - Matches, and captures, ", ', or empty string
							// \s             - Ensures there's a space after the attribute
							$new_tag = preg_replace( '#(?<=\s)(width=["\']?)[\d%]+(["\']?)\s?#i', sprintf( '${1}%d${2} ', $resize_args[0] ), $new_tag );
							$new_tag = preg_replace( '#(?<=\s)(height=["\']?)[\d%]+(["\']?)\s?#i', sprintf( '${1}%d${2} ', $resize_args[1] ), $new_tag );
```

anomiex commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -0,0 +1,1386 @@
+						$content = str_replace( $tag, $new_tag, $content );
+					}
+				} elseif ( preg_match( '#^http(s)?://i[\d]{1}.wp.com#', $src ) && ! empty( $images['link_url'][ $index ] ) && self::validate_image_url( $images['link_url'][ $index ] ) ) {
+					$new_tag = preg_replace( '#(href=["|\'])' . preg_quote( $images['link_url'][ $index ], '#' ) . '(["|\'])#i', '\1' . Image_CDN_Core::cdn_url( $images['link_url'][ $index ] ) . '\2', $tag, 1 );
```
```suggestion
					$new_tag = preg_replace( '#(href=["\'])' . preg_quote( $images['link_url'][ $index ], '#' ) . '(["\'])#i', '\1' . Image_CDN_Core::cdn_url( $images['link_url'][ $index ] ) . '\2', $tag, 1 );
```

anomiex commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -0,0 +1,1386 @@
+				continue;
+			}
+
+			/** This filter is already documented in class.photon.php */
```
```suggestion
			/** This filter is already documented in class-image-cdn.php */
```
?

anomiex commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -0,0 +1,1386 @@
+		if (
+			/** Short-circuit via jetpack_photon_srcset_multipliers filter. */
+			is_array( $multipliers )
+			/** This filter is already documented in class.photon.php */
```
```suggestion
			/** This filter is already documented in class-image-cdn.php */
```
?

anomiex commented on projects/packages/image-cdn/src/class-image-cdn.php:
```diff
@@ -0,0 +1,1386 @@
+			&& isset( $image_meta['width'] ) && isset( $image_meta['height'] ) && isset( $image_meta['file'] )
+			/** Verify we have the requested width/height. */
+			&& isset( $size_array[0] ) && isset( $size_array[1] )
+			) {
```
Nitpick
```suggestion
		) {
```

anomiex reviewed (commented):
Gave the code a once-over, picked a few nits and flagged a few potential issues.

thingalon committed:
Minor correction

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

thingalon committed:
Formatting change

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

thingalon committed:
Grammar fix

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

thingalon committed:
Use `self` in preference to naming the class again

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

thingalon committed:
Cleanup spacing

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

thingalon committed:
Correct reference to photon in a comment

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

thingalon committed:
Correct photon ref

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

haqadn committed:
Fix class reference

haqadn committed:
Update docblock references

haqadn commented:
Thank you, @anomiex, for being thorough and identifying the issues. While I have addressed some of them, and @thingalon has addressed a few more, I feel we should address the preexisting bug fixes in separate PRs to keep the intention of this PR clear and unblock a few other things. I have made a note in the follow up github issue to address the issues you raised #29577

Commit:
Defensive code: remove Jetpack_Photon verison of filter if present

Commit:
Make image_sizes private

Commit:
Finalize Image_CDN class

Commit:
remove | as a valid quotation character

Commit:
Remove since tags from methods which now live in a new class

thingalon committed:
Update projects/packages/image-cdn/src/class-image-cdn.php

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

thingalon committed:
Update projects/packages/image-cdn/src/class-image-cdn.php

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

thingalon committed:
Update projects/packages/image-cdn/src/class-image-cdn.php

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

thingalon committed:
Update projects/packages/image-cdn/src/class-image-cdn.php

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

thingalon committed:
Update projects/packages/image-cdn/src/class-image-cdn.php

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

thingalon committed:
Update projects/packages/image-cdn/src/class-image-cdn.php

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

thingalon committed:
Update projects/packages/image-cdn/src/class-image-cdn.php

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

thingalon committed:
Update projects/packages/image-cdn/src/class-image-cdn.php

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

thingalon committed:
Update projects/packages/image-cdn/src/class-image-cdn.php

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

haqadn committed:
Move photon compatibility to image-cdn

haqadn committed:
Update readme

thingalon reviewed (approved):
This is looking great to me 🚀

thingalon commented:
IMHO this is good to go. In particular, hat tip to @anomiex for such a thorough review.

Is there a guide for us setting up a production repository for the new module, or is this something w should ask @Automattic/ground-control  or @Automattic/jetpack-crew to handle?

jeherve commented:
> Is there a guide for us setting up a production repository for the new module

It's ready for you, I set it up when you created this PR last week. All you have to do is merge this PR.

- https://github.com/Automattic/jetpack-image-cdn
- https://packagist.org/packages/automattic/jetpack-image-cdn

anomiex commented:
> Is there a guide for us setting up a production repository for the new module, or is this something w should ask @Automattic/ground-control or @Automattic/jetpack-crew to handle?

There are docs at https://github.com/Automattic/jetpack/blob/trunk/docs/monorepo.md#mirror-repositories, but we're happy to handle it for you (like Jeremy already did for this one).
